pyogrio/util.py

```diff
@@ -0,0 +1,113 @@
+"""Dataset paths, identifiers, and filenames"""
+
+import re
+import sys
+from urllib.parse import urlparse
+
+
+def vsi_path(path: str) -> str:
+    """
+    Ensure path is a local path or a GDAL-compatible vsi path.
+
+    """
+    # Windows drive letters (e.g. "C:\") confuse `urlparse` as they look like
+    # URL schemes
+    if sys.platform == "win32" and re.match("^[a-zA-Z]\\:", path):
+        return path
+
+    elif path.startswith("/vsi"):
+        return path
+
+    elif re.match("^[a-z0-9\\+]*://", path):
+
+        path, archive, scheme = _parse_uri(path)
+        if not scheme:
+            return path
+
+        return _construct_vsi_path(path, archive, scheme)
+
+    else:
+        return path
+
+
+# Supported URI schemes and their mapping to GDAL's VSI suffix.
+SCHEMES = {
+    "ftp": "curl",
+    "gzip": "gzip",
+    "http": "curl",
+    "https": "curl",
+    "s3": "s3",
+    "tar": "tar",
+    "zip": "zip",
+    "file": "file",
+    "gs": "gs",
+    "az": "az",
+    "adls": "adls",
+    "adl": "adls",  # fsspec uses this
+    # 'oss': 'oss',
```

Comment:
Please remove the commented lines or add a TODO / GH issue.

Can you please add tests for `hdfs` and `webhdfs`?

Reply:
Added test for hdfs/webhdfs. 

The reason I commented out "oss" and "swift" for now (while those are supported by GDAL) is that I don't actually know if those are used as URI in practice (eg I am not aware that you can use a "oss://path/data" URI in Python with an fsspec implementation, and in general haven't found any example online based on a quick google search). So I am a bit in doubt if we should actually enable those (the user can always still use the `/vsioss/..` path as well)

Reply:
Ok - seems like a good idea to leave them out here; having them here implies that we support them.  Can always add them later when there is a clear need.